phoenix-core-client/src/main/java/org/apache/phoenix/expression/util/bson/SQLComparisonExpressionUtils.java

```diff
@@ -0,0 +1,290 @@
+  private static final Pattern FIELD_EXISTS_PATTERN = Pattern.compile(FIELD_EXISTS);
+  private static final Pattern EQUALS1_PATTERN = Pattern.compile(EQUALS1);
+  private static final Pattern EQUALS2_PATTERN = Pattern.compile(EQUALS2);
+  private static final Pattern NOT_EQUALS1_PATTERN = Pattern.compile(NOT_EQUALS1);
+  private static final Pattern NOT_EQUALS2_PATTERN = Pattern.compile(NOT_EQUALS2);
+  private static final Pattern LESS_THAN_PATTERN = Pattern.compile(LESS_THAN);
+  private static final Pattern LESS_THAN_OR_EQUALS_PATTERN = Pattern.compile(LESS_THAN_OR_EQUALS);
+  private static final Pattern GREATER_THAN_PATTERN = Pattern.compile(GREATER_THAN);
+  private static final Pattern GREATER_THAN_OR_EQUALS_PATTERN =
+          Pattern.compile(GREATER_THAN_OR_EQUALS);
+  private static final Pattern BETWEEN_PATTERN = Pattern.compile(BETWEEN);
+  private static final Pattern IN_PATTERN = Pattern.compile(IN);
+  private static final Pattern AND_PATTERN = Pattern.compile(AND);
+  private static final Pattern OR_PATTERN = Pattern.compile(OR);
+  private static final Pattern NOT_PATTERN = Pattern.compile(NOT);
+
+  private static final String FUNC_FIELD_NOT_EXISTS = "!exists('$1')";
+  private static final String FUNC_FIELD_EXISTS = "exists('$1')";
+  private static final String FUNC_EQUALS = "isEquals('$1', '$2')";
+  private static final String FUNC_NOT_EQUALS = "!isEquals('$1', '$2')";
+  private static final String FUNC_LESS_THAN = "lessThan('$1', '$2')";
+  private static final String FUNC_LESS_THAN_OR_EQUALS = "lessThanOrEquals('$1', '$2')";
+  private static final String FUNC_GREATER_THAN = "greaterThan('$1', '$2')";
+  private static final String FUNC_GREATER_THAN_OR_EQUALS = "greaterThanOrEquals('$1', '$2')";
+  private static final String FUNC_BETWEEN = "between('$1', '$2', '$3')";
+  private static final String FUNC_IN = "in('$1', '$2')";
+  private static final String OP_AND = "&&";
+  private static final String OP_OR = "||";
+  private static final String OP_NOT = "!";
+
+  private final RawBsonDocument rawBsonDocument;
+  private final BsonDocument comparisonValuesDocument;
+
+  public SQLComparisonExpressionUtils(RawBsonDocument rawBsonDocument,
+      BsonDocument comparisonValuesDocument) {
+    this.rawBsonDocument = rawBsonDocument;
+    this.comparisonValuesDocument = comparisonValuesDocument;
+  }
+
+  public boolean isConditionExpressionMatching(final String conditionExpression) {
+    String expression = convertExpression(conditionExpression);
+    LOGGER.info("Evaluating Expression: {}", expression);
+    try {
+      Object result = MVEL.eval(expression, this);
+      return result instanceof Boolean && (Boolean) result;
+    } catch (Exception e) {
+      LOGGER.error("Error while evaluating expression: {}", expression, e);
+      throw new RuntimeException("Expression could not be evaluated: " + expression);
+    }
+  }
+
+  public String convertExpression(String expression) {
```

Comment:
This method scans an expression string 15 times for pattern matching and then 15 times for replacing matching strings. A conventional parsing would need only one pass.  Also it creates 15 matcher objects and 15 new expressions via this process. This looks computationally pretty expensive.

Reply:
Yes, the string is parsed 15 times for pattern matching and this could be a bit time consuming and I have been exploring a couple of options:

1. Using Pattern matcher replacement with StringBuffer
2. SearchAndReplaceBuilder from DFALEX: https://github.com/mtimmerm/dfalex/commit/a1bc6aca051b7b2cbdc7e2c9e639fe8a1e3cd16a
3. commons-text StringSubstitutor

None of these options are exactly helping with what we want.

I have also been measuring as large string as i could generate. With all the strings available within tests, the max time spent in the computation is 1ms (it fluctuates b/ 0 and 1 ms).

I am still exploring more options but so far have not seen more efficient multi-pattern matcher for single string in Java.

Reply:
Also, earlier i was doing String#replaceAll, which would compile regex pattern each time string replacement happens. With the latest commit, i have separated out pattern compilation and kept it static so it happens only once: https://github.com/apache/phoenix/pull/1906/commits/ace8240f1df4c027ce2458309f8ed78c05f5484b

Reply:
Updated PR to reduce the num of pattern matching and replacement. The pattern matching with replacement will be in direct correlation with the num of unique operators used in the String.

Reply:
Moreover, to further reduce the dependency on the pattern-matcher, we can also document option for client to use executable functions directly i.e. instead of `a = :b`, client can directly provide `isEquals('a', ':b')`, similarly instead of `AND`, client can provide `&&` etc, in which case we will no longer perform pattern-matcher-replacer for the input string.

Reply:
In the latest revision, I just added additional tests that no longer relies on pattern-matcher for each matcher replacement, as the string input itself is executable.

Reply:
I had an offline discussion with @kadirozde reg this. I will create a separate follow-up Jira to use ANTLR based parsing for Bson SQL style condition expression evaluation.